PEP 1: Update for governance changes

ncoghlan opened the pull request:
NOTE: Currently posted for Steering Council review, so definitely don't merge this yet :)

Minimalist adjustment to the PEP process to account for the
language governance changes.

- final design authority is now the elected Steering Council
- BDFL-Delegate process is now used for all PEPs, with the
  Steering Council looking to appoint a delegate if nobody
  volunteers of their own accord, and PEPs being deferred
  if no suitable delegate can be found
- avoiding the role renaming bikeshed for the time being by
  keeping the BDFL-Delegate title

This will allow us to start processing additional PEPs for
Python 3.8.

I also addressed some existing issues that the above changes
made more readily apparent:

- clarified that the PEP process is used for more than just
  the CPython reference interpreter
- clarified the "standing delegation" model that we have in
  place for packaging related PEPs

willingc commented on pep-0001.txt:
```diff
@@ -230,28 +238,40 @@ PEP Review & Resolution
-before the PEP can be accepted.
+the Steering Council's delegate for that PEP.
+
+If their self-nomination is accepted by the other core developers and the
```
Perhaps:

...self-nomination has broad consensus of the other core developers and the acceptance of the Steering Council, they ...

warsaw replied:
Perhaps we should turn that around, and instead of "broad consensus", operate on a veto basis.  E.g. someone self-nominates for BDFL-delegate.  Other core devs can petition the SC to veto that delegation.  If they concur, then a new delegate must be found.  If not, then the original delegate is approved.  I.e. only the SC can veto a delegate self-nomination, but core devs can petition the SC for that ruling if they think something is amiss.

ncoghlan replied:
I almost changed this to assume consensus in the absence of a Steering Council veto to bring it closer to the wording in https://www.python.org/dev/peps/pep-0013/#membership (just without the complexity of core devs having to formally vote on every BDFL-Delegate appointment), and only left it alone on the basis of minimalism.

I'll go ahead and make that change now, since it isn't just me that found the current wording underspecified :)

warsaw commented on pep-0001.txt:
```diff
@@ -230,28 +238,40 @@ PEP Review & Resolution
-predetermined to be acceptable (e.g., it is an obvious win as-is
-and/or its implementation has already been checked in) the BDFL may
-also initiate a PEP review, first notifying the PEP author(s) and
+final acceptance of the PEP must be requested of the core developers, usually
```
How would this work?  As I read the change, the core developers (not the SC or their delegate) can accept a PEP, but I don't think that process has been worked out yet.  The SC has yet to "establish appropriate decision-making processes for PEPs" as quoted from [PEP 13](https://www.python.org/dev/peps/pep-0013/).

ncoghlan replied:
I'm trying to word this part in a way that *doesn't* encourage people to send PEP acceptance requests to the steering council mailing list.

As far as how the core devs collectively accept PEPs goes:

1. Core devs collectively elect a Steering Council
2. Steering Council appoints (or chooses not to veto) BDFL-Delegates (who will be core developers themselves in all but the most extraordinary cases)
3. BDFL-Delegates accept or reject individual PEPs

ncoghlan replied:
Ahah, I have a cunning plan: delete the sentence. It adds no useful information, as it was just restating what the rest of this section explains, but restating it badly :)

warsaw commented on pep-0001.txt:
```diff
@@ -230,28 +238,40 @@ PEP Review & Resolution
+
+For a PEP that is predetermined to be acceptable (e.g., it is an obvious win
+as-is and/or its implementation has already been checked in) the Steering
+Council may also initiate a PEP review, first notifying the PEP author(s) and
```
Hmm, why only for a predetermined PEP?  And if it is predetermined, then what is the purpose of the SC review?

ncoghlan replied:
I was wondering why this was showing as a change, and then realised it was because I broke it up into two paragraphs.

I'll reword this to just say that the SC can initiate reviews, without the editorialising on why they might want to do that.

ncoghlan replied:
And in attempting to do that, I realised why the edititorialising was there: without it, the sentence left me wondering why the SC would ever do that, given that the cost of leaving Open PEPs languishing is pretty low.

And the main reason for the SC to do this is because they want to *accept* the PEP, so the implementation can be merged before a release. Anyway, new wording coming in the next push.

warsaw commented on pep-0001.txt:
```diff
@@ -469,10 +489,10 @@ following RFC 2822 continuation line conventions.  Note that personal
-approve or reject a PEP rests with someone other than the BDFL. (The
-delegate's email address is currently omitted due to a limitation in the
-email address masking for reStructuredText PEPs)
+The BDFL-Delegate field is used to record the core developer appointed by the
```
Do we want to limit BDFL-delegate position to core devs only?  If so, do we say that explicitly?  To me, this implies that but doesn't make it explicit.  What was your intention?

willingc replied:
core developer -> person

Given that there are PEPs that deal with workflow, etc., it seems reasonable not to overly limit the position to core dev.

ncoghlan replied:
We also use the PEP process for non-CPython activities (e.g. PyPA). Hmm, I should make that clearer in the early parts of the PEP - at the moment it's buried in an aside down in the body of the PEP about the `Discussions-To` header...

ncoghlan replied:
'person' didn't quite scan right for me, so I went with 'individual'.

ncoghlan commented on pep-0001.txt:
```diff
@@ -62,16 +62,24 @@ There are three kinds of PEP:
+van Rossum, the original creator of the Python programming language.  While
+"BDFL" turned out to be the more pragmatic "BDUHSD" ("Benevolent Dictator
+Until He Stepped Down") in practice, the previous title has been retained
+for the time being.
```
Will do.

ncoghlan replied:
I wanted to at least acknowledge the change in governance model that created the historical term, so my latest revision here covers the source of design authority in the two models.

The paragraph is about you though, so it probably makes the most sense for you to edit it directly to something that you like (everyone should have edit access on the PR).

gvanrossum replied:
What yo did is fine!

ncoghlan commented:
Updated based on the review feedback, and to address a couple of additional problems I noticed around the standing delegations to PyPA for packaging interoperability PEPs.

I don't think the latter issue is actually fully fixed yet (it's weird that there's no link out to https://www.pypa.io/en/latest/specifications/#proposing-new-specifications from PEP 1, making the PyPA page the only record of the standing delegations), but fixing it properly should be a separate discussion that includes Paul and Donald, and it isn't a blocker for getting Python 3.8 PEPs moving again, so it doesn't need to be resolved here.

willingc commented on pep-0001.txt:
```diff
@@ -273,11 +297,19 @@ down by the Council, in which case a new BDFL-Delegate will be appointed in the
+that don't immediately affect the standard library). In this case, the
+"Discussions-To" heading in the PEP will identify the appropriate alternative
+list where discussion, review and pronouncement on the PEP will occur. The
+Steering Council may also choose to set up "standing delegations" for these
```
Perhaps we should remove the standing delegation wording. To me, this is more an implementation detail than governance. For example, the Steering Council could, without additional wording here, delegate all packaging PEPs to one or two individuals until the end of the Steering Council term.

warsaw replied:
How will we record the setting up of these delegation?  As @willingc says, the SC is within its authority to delegate that responsibility until the end of its term.  However *at* the end of the term, a potentially entirely new SC will come in.  They should have a definitive list of those delegations, which they could simply continue with an explicit declaration.  But that would be difficult if they didn't have that definitive list of delegations.

willingc replied:
@warsaw I think that the list could be maintained in the new "steering-council" repo which we will move from my account to python. In addition to a markdown list, it would also likely be captured in biweekly reports.

ncoghlan replied:
Yeah, this is what I was referring to when I commented that I didn't think what I had here was actually enough to fix the gaps in the current standing delegation process, since we're still missing a record on the Python level governance side of the standing delegations that I listed in the PyPA specification management guidelines.

Between Carol's and Barry's comment at https://github.com/python/peps/pull/896#discussion_r257521315 though, I think this paragraph will be improved by making it *less* specific, and instead just reminding the Steering Council members that if we set something like this up, we should write it down somewhere where the next Steering Council can find it: https://github.com/python/peps/pull/896/commits/f851ef54bf6a4b8bf61f4314c7054be11d99f450

So the new wording just says that there should *be* records, but it doesn't attempt to specify what form those records will take.

warsaw commented on pep-0001.txt:
```diff
@@ -229,45 +255,71 @@ PEP Review & Resolution
+replacement can be found).
+
+When the PEP process is being used for projects other than the CPython
+reference interpreter, PEP review and resolution may also occur on a list
```
I would like to explicitly include "Python the language" here.  IOW from:

"other than the CPython reference interpreter"

to

"other than the Python language or CPython reference implementation"

Maybe that's obvious, but I think it would be useful to be clear that discussions about the language irrespective of the implementation, also belong on python-dev.

ncoghlan replied:
I ended up going the other direction, and just saying that cases like this require explicit Steering Council approval, rather than attempting to enumerate the situations where the Steering Council probably *wouldn't* give approval.

ncoghlan commented:
Latest update addresses comments from @warsaw and @willingc  by reducing the amount of detail in the PEP itself regarding standing delegations, and instead just makes it clear that:

1. the Steering Council has the authority to set them up;
2. when we do, we should document them properly

Once there's a specific location for those records in the python/steering-council repo it will make sense to come back to this and add a direct link, but for now, the statement of intent is enough to address the concern I had with the old wording.

ncoghlan commented:
Also noting something I deliberately chose to leave alone for now: the PEP as a whole is mailing list centric, and would need some tweaks to properly account for the possibility of using other formats (e.g. web forums, issue trackers) for discussions and decision making.

I figure rather than attempting to pre-empt that need, it makes more sense to wait until specific situation comes up to give a clearer picture of the exact changes that would be needed to cover that.

gvanrossum commented on pep-0001.txt:
```diff
@@ -229,45 +255,69 @@ PEP Review & Resolution
+occur.
+
+When such standing delegations are put in place, the Steering Council will
+maintain sufficient records to allow subsequent Councils to understand the
```
Maybe the core devs and the general public should also be audiences of these records?

willingc replied:
Yes!

ncoghlan replied:
Revised wording makes it explicit that these records should be public and for a general audience: https://github.com/python/peps/pull/896/commits/a50dbc7e1e69341645d27851c8a202f132fff05d

(I hadn't even noticed that the original wording technically allowed those records to be "for Steering Council eyes only")

gvanrossum commented on pep-0001.txt:
```diff
@@ -229,45 +255,69 @@ PEP Review & Resolution
+BDFL-Delegate for that PEP. If no suitable candidate can be found, then the
+PEP will be marked as Deferred until one is available.
+
+While BDFL-Delegates hold final authority over acceptance or rejection of a PEP,
```
I wonder if we should ratification by the SC in the PEP approval process. I could just see a rogue BDFL-Delegate announcing approval of a PEP before they are asked to step down by the SC (or perhaps even refusing to step down). It would be nice if the PEP made clear that in this case the SC wins. (If the SC itself has gone rogue, PEP 13 has ways to deal with that.)

warsaw replied:
+1 - I don't think it's too much of a burden for the SC to ratify.  It'll usually be rubber stamped but I think it does give a little extra legitimacy and checks-n-balances on the process.

willingc replied:
+1

ncoghlan replied:
I agree the handling of this case needs to be clearer.

In https://github.com/python/peps/pull/896/commits/bad24f3c57dfb9f97a62d931c2f058b3db3166c0, rather than adding a ratification step, I instead made it so that if a BDFL-Delegate is asked to stand down by the Council, any prior decision about acceptance or rejection is reverted (and that reversion is triggered by the request itself, as if the SC were happy with the decision that had already been made, they wouldn't be asking the delegate to stand down).

If folks have a strong preference for routine ratification, then I can find a way to reword that section. However, I do think going down that path would mostly just be annoying for BDFL-Delegates and Steering Council members - in a work context, I've been in the position of both seeking the rubberstamp and of wielding it, and it irks me no matter which side I'm on.

ncoghlan commented:
Reviewing the PR myself, I caught an issue in the new `PEP Audience` section, which is that I had omitted developers working on implementations other than CPython: https://github.com/python/peps/pull/896/commits/b531a9d20a2135258c7068b8c8aa262b2f885244

gvanrossum commented:
Thanks for the care and thought put into this PR, Nick!

brettcannon commented:
Since I was the final hold up on providing approval (sorry about that; busy week), I went ahead and merged this.

Thanks for writing up the changes, @ncoghlan !

ncoghlan commented:
Thanks for the excellent reviews! The merged version is much better than my initial draft :)
